src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs

```diff
@@ -83,22 +83,26 @@ public static int IndexOf(ref byte searchSpace, int searchSpaceLength, ref byte
                 {
                     Vector256<byte> cmpCh1 = Vector256.Equals(ch1, Vector256.LoadUnsafe(ref searchSpace, (nuint)offset));
                     Vector256<byte> cmpCh2 = Vector256.Equals(ch2, Vector256.LoadUnsafe(ref searchSpace, (nuint)(offset + ch1ch2Distance)));
-                    uint mask = (cmpCh1 & cmpCh2).AsByte().ExtractMostSignificantBits();
+                    Vector256<byte> cmpAnd = (cmpCh1 & cmpCh2).AsByte();
 
-                    while (mask != 0)
+                    if (!cmpAnd.IsAllZero())
                     {
-                        int bitPos = BitOperations.TrailingZeroCount(mask);
-                        if (valueTailNLength == 1 || // we already matched two bytes
-                            SequenceEqual(
-                                ref Unsafe.Add(ref searchSpace, offset + bitPos + 1),
-                                ref valueTail,
-                                valueTailNLength))
+                        uint mask = cmpAnd.ExtractMostSignificantBits();
+                        while (mask != 0)
                         {
-                            return offset + bitPos;
-                        }
+                            int bitPos = BitOperations.TrailingZeroCount(mask);
+                            if (valueTailNLength == 1 || // we already matched two bytes
+                                SequenceEqual(
+                                    ref Unsafe.Add(ref searchSpace, offset + bitPos + 1),
+                                    ref valueTail,
+                                    valueTailNLength))
+                            {
+                                return offset + bitPos;
+                            }
 
-                        // Clear lowest set bit
-                        mask = Bmi1.IsSupported ? Bmi1.ResetLowestSetBit(mask) : mask & (mask - 1);
+                            // Clear lowest set bit
+                            mask = Bmi1.IsSupported ? Bmi1.ResetLowestSetBit(mask) : mask & (mask - 1);
```

Comment:
This is potentially a good API to propose for `BitOperations`. Likewise there is a community PR to automatically recognize the appropriate pattern(s) and convert it to `bslr`, so I'd opt for either a private helper on BitOperations + API proposal; or just using the simple pattern and letting the JIT optimize it.